docs/connections.md

```diff
@@ -0,0 +1,597 @@
+appeared in any non-root code block.
+
+## Lambdas
+`arg -> value` is the syntax for lambdas. Left-hand side is a pattern for the
+argument (lambdas are always unary) and the right-hand side is its body.
+Right-hand side, i.e. the lambda body, introduces its own scope.
+
+If the lambda is *not* introduced in what is already a pattern context, the `->`'s pattern introduces identifiers into the scope of the right-hand
+side.
+
+Example:
+```
+«succ» = a -> a + 1
+«foo» = »a«
+```
+
+Here lambda introduces `a` only into its right-hand side. The `a` that is being
+assigned to `foo` is not the same `a` as in lambda — it must be defined
+elsewhere in the module or the code will fail.
+
+However, if `->` appears in a pattern context, its left-hand side identifiers
+are introduced into the scope targeted by the outer pattern context.
+
+Example:
+```
+(a -> b) -> a.»default«
+```
+
+If not for this second rule, the `a` would be visible only in place of
+expression `b`. However, now it is visible in the outer lambda body and can be
+accessed. The only externally provided identifier must be `default` method.
+
+--- 
+Lambdas may not appear in the assignment's pattern (i.e. values cannot be
+pattern-matched into lambda).
+
+So the following is not valid:
+```
+a -> b = foo
+```
+
+---
+Example:
+```
+a -> a -> b
+```
+
+Here the first `a` and second `a` are separate identifiers, the latter shadowing
+the first one. If one wanted to express that both arguments are of the same
+type, `a -> A -> b` would have been used. `b` refers to an identifier from
+graph's scope (it is in the body's position, not pattern).
+
+OPEN QUESTION: actually it might be "nice" to have both `a` unified in such
+case. 
+
+
+
+## Type ascription
+The type ascription operator `:` introduces pattern context for its right hand
+side. The basic form is `value:type`. It says that `value` be of the given
+`type`, i.e. that all its possible values belong to the set of atoms represented
+by type.
+
+The effect of this can be two-fold. If `value` is of (at least partially) known
+type, appropriate constraints will be introduced on the types denoted by
+variable names appearing in the pattern context. If the identifier was not
+defined, it will appear in the current scope.
+
+For example:
+```
+5 : «a»?
+```
+
+This introduces constraint on the type `a` that its value set must include atom
+`5`. If the `a` is already visible in the scope, this constraint will be added.
+Otherwise, `a` will be introduced into the current scope with that constraint.
+
+// TODO: What if parent scope only ascribes identifier with type constraint but
+only nested scope assigns to it? 
+
+// TODO: Open design question, if `a` should modify existing variable or should
+always try to shadow it. What is the difference between `5 : a` and `5 : A`?
+(except the latter not being able to ever introduce a new identifier)
+
+// TODO: Is it legal or sensible to ascribe a variable when there is no binding?
```

Comment:
I don't think it should be legal. A type signature should always be associated with a binding for that name.